### PR TITLE
Pre-commit and ruff configuration

### LINKED ISSUE #204: Integrate Ruff Linter for Enhanced Code Quality in Python Project
# Overview
This issue aims to enhance our Python project's code quality by integrating the Ruff linter and setting up a pre-commit hook to ensure linting checks are run automatically before each commit.

# Goals
- Integrate Ruff Linter: Add the Ruff linter to our project's development environment.
- Configure Pre-Commit Hook: Implement a pre-commit hook that runs the Ruff linter against staged files before each commit.
- Ensure Seamless Integration: Verify that the integration of Ruff and the pre-commit hook doesn't disrupt the development workflow and ensures consistent code quality.

# Tasks
- [x] Research and install Ruff linter.
- [x] Update project dependencies and configuration to include Ruff.
- [x] Run the linter on all existing files and fix issues
- [x] Configure pre-commit hook to run Ruff against staged files.
- [x] Test the pre-commit hook to ensure it successfully runs the Ruff linter.
- [ ] Test
- [ ] Add documentation

### PR DESCRIPTION
**Pull request by DevilsAutumn**
closes #204

### CODE REVIEW

**Comment by deepakdinesh1123** [comment]
@DevilsAutumn can you check if there is a pre-commit hook or any way to make sure that secrets and keys are not committed?

**Comment by DevilsAutumn** [comment]
> @DevilsAutumn can you check if there is a pre-commit hook or any way to make sure that secrets and keys are not committed?

Added https://github.com/pre-commit/pre-commit-hooks?tab=readme-ov-file#detect-private-key

**Comment by kc-diabeat** [comment]
@DevilsAutumn How are we installing the pre-commit package? Don't see it in the requirements.txt file?

**Comment by DevilsAutumn** [comment]
> @DevilsAutumn How are we installing the pre-commit package? Don't see it in the requirements.txt file?

My bad, I forgot to add it in requirements.txt. Added now.

File: backend/src/zango/apps/accesslogs/apps.py [hunk 1]
```diff
@@ -2,8 +2,7 @@
 
 
 class AccesslogsAppConfig(AppConfig):
-
     name = "zango.apps.accesslogs"
 
     def ready(self):
-        import zango.apps.accesslogs.signals
```
Comment (shahharsh176) [inline backend/src/zango/apps/accesslogs/apps.py:9]: @DevilsAutumn 
importing signals will be required

please check it in other places also

Comment (DevilsAutumn) [reply backend/src/zango/apps/accesslogs/apps.py:9]: okay

**Comment by deepakdinesh1123** [comment]
@DevilsAutumn can we use [Interrogate](https://interrogate.readthedocs.io/en/latest/index.html) to check if docstrings are added to methods?

**Comment by DevilsAutumn** [comment]
> @DevilsAutumn can we use [Interrogate](https://interrogate.readthedocs.io/en/latest/index.html) to check if docstrings are added to methods?

Sure!

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  hooks:
+  - id: check-yaml
+  - id: check-case-conflict
+  - id: check-json
+  - id: check-merge-conflict
+  - id: check-symlinks
+  - id: check-toml
+  - id: end-of-file-fixer
+  - id: requirements-txt-fixer
+  - id: detect-private-key
```
Comment (deepakdinesh1123) [inline .pre-commit-config.yaml:15]: please add the hook `detect-aws-credentials`

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+- repo: https://github.com/astral-sh/ruff-pre-commit
```
Comment (deepakdinesh1123) [inline .pre-commit-config.yaml:16]: Also add this hook to detect secrets

```yaml
-   repo: https://github.com/Yelp/detect-secrets
    rev: v1.5.0
    hooks:
    -   id: detect-secrets
        exclude: package.lock.json
```

**Review by deepakdinesh1123 — CHANGES_REQUESTED** [review]
@DevilsAutumn I have requested changed related to scanning for secrets please take a look at them

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  rev: v0.3.5
+  hooks:
+    - id: ruff
+      args: [ --fix ]
+    - id: ruff-format
+
```
Comment (deepakdinesh1123) [inline .pre-commit-config.yaml:22]: please configure isort to sort the inputs in this order

- built in packages
- third party python packages
- django imports
- zango imports
- relative imports

**Review by deepakdinesh1123 — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+- repo: https://github.com/pre-commit/pre-commit-hooks
+  rev: v4.5.0
+  hooks:
+  - id: check-yaml
+  - id: check-case-conflict
+  - id: check-json
+  - id: check-merge-conflict
+  - id: check-symlinks
+  - id: check-toml
+  - id: end-of-file-fixer
```
Comment (deepakdinesh1123) [inline .pre-commit-config.yaml:13]: when running pre-commit locally I saw that it was fixing files that did not have a newline at the end, but in the review saw that some of the files did not have a newline, please run `pre-commit run --all-files` once and commit the results

File: backend/src/zango/apps/shared/platformauth/apps.py [hunk 6]
```diff
@@ -7,4 +7,4 @@ class AuthConfig(AppConfig):
 
     def ready(self):
         # Import the models here
-        from .models import PlatformUserModel
```
Comment (deepakdinesh1123) [inline backend/src/zango/apps/shared/platformauth/apps.py:10]: please add this import

**Review by deepakdinesh1123 — CHANGES_REQUESTED** [review]

**Comment by kc-diabeat** [comment]
The build artifact is available in the artifact section [here](https://github.com/Healthlane-Technologies/Zango/actions/runs/10560048582). You can download it and place the build file under backend/src/zango/assets/app_panel/js for local testing and verification.

**Comment by deepakdinesh1123** [comment]
@DevilsAutumn Please exclude migrations folder from pre-commit and add docs on how to setup and use pre-commit in the contribution doc

**Comment by kc-diabeat** [comment]
The build artifact is available in the artifact section [here](https://github.com/Healthlane-Technologies/Zango/actions/runs/10573647101). You can download it and place the build file under backend/src/zango/assets/app_panel/js for local testing and verification.

**Review by deepakdinesh1123 — APPROVED** [review]

**Comment by shahharsh176** [comment]
@DevilsAutumn Can we revert the changes made in frontend folder ?

**Comment by kc-diabeat** [comment]
The build artifact is available in the artifact section [here](https://github.com/Healthlane-Technologies/Zango/actions/runs/10578360213). You can download it and place the build file under backend/src/zango/assets/app_panel/js for local testing and verification.

**Comment by kc-diabeat** [comment]
The build artifact is available in the artifact section [here](https://github.com/Healthlane-Technologies/Zango/actions/runs/10578408239). You can download it and place the build file under backend/src/zango/assets/app_panel/js for local testing and verification.

**Comment by DevilsAutumn** [comment]
> @DevilsAutumn Can we revert the changes made in frontend folder ?

Done!

File: backend/src/zango/apps/auditlogs/registry.py [hunk 7]
```diff
@@ -47,7 +42,7 @@ def __init__(
         delete: bool = True,
         access: bool = True,
         m2m: bool = True,
-        custom: Optional[Dict[ModelSignal, Callable]] = None,
+        custom: dict[ModelSignal, Callable] | None = None,
```
Comment (shahharsh176) [inline backend/src/zango/apps/auditlogs/registry.py:45]: This is raising error in python 3.9.6
```
custom: dict[ModelSignal, Callable] | None = None,
TypeError: unsupported operand type(s) for |: 'types.GenericAlias' and 'NoneType'
```

Comment (DevilsAutumn) [reply backend/src/zango/apps/auditlogs/registry.py]: Fixed, Should be working now.

File: backend/src/zango/apps/auditlogs/registry.py [hunk 8]
```diff
@@ -264,7 +257,7 @@ def _get_exclude_models(
         ]
         return exclude_models
 
-    def _register_models(self, models: Iterable[Union[str, Dict[str, Any]]]) -> None:
+    def _register_models(self, models: Iterable[str | dict[str, Any]]) -> None:
```
Comment (shahharsh176) [inline backend/src/zango/apps/auditlogs/registry.py:260]: @DevilsAutumn This is also breaking in python 3.9.6
```
    class AuditlogModelRegistry:
  File "/Users/harshshah/Documents/Zelthy_DEV/Zelthy-V2/zel3_main_dev/zelthy3/backend/src/zango/apps/auditlogs/registry.py", line 260, in AuditlogModelRegistry
    def _register_models(self, models: Iterable[str | dict[str, Any]]) -> None:
TypeError: unsupported operand type(s) for |: 'type' and 'types.GenericAlias'
```

Comment (DevilsAutumn) [reply backend/src/zango/apps/auditlogs/registry.py]: fixed.

File: backend/src/zango/apps/permissions/models.py [hunk 9]
```diff
@@ -1,10 +1,10 @@
-import uuid
-from django.db.models import Q
+from time import timezone
```
Comment (shahharsh176) [inline backend/src/zango/apps/permissions/models.py:1]: @DevilsAutumn Can you pls change this to `from django.utils import timezone` ?

File: backend/src/zango/assets/app_panel/js/build.1723136204750.min.js
Comment (shahharsh176) [inline backend/src/zango/assets/app_panel/js/build.1723136204750.min.js:1]: Can you please replace this with build present in main branch ?

File: backend/src/zango/assets/js/jquery/3.7.1/jquery.min.js
Comment (shahharsh176) [inline backend/src/zango/assets/js/jquery/3.7.1/jquery.min.js:1]: I think this diff is because of EOL formatted right ? Please verify it once

**Review by shahharsh176 — APPROVED** [review]

**Review by kc-diabeat — APPROVED** [review]

**Pull request merged**